### PR TITLE
[SPARK-28160][CORE] Fix a bug that callback function may hang when unchecked exception missed

### PR DESCRIPTION
**Pull request by LantaoJin**
## What changes were proposed in this pull request?

This is very like #23590 .

`ByteBuffer.allocate` may throw `OutOfMemoryError` when the response is large but no enough memory is available. However, when this happens, `TransportClient.sendRpcSync` will just hang forever if the timeout set to unlimited.

This PR catches `Throwable` and uses the error to complete `SettableFuture`.

## How was this patch tested?

I tested in my IDE by setting the value of size to -1 to verify the result. Without this patch, it won't be finished until timeout (May hang forever if timeout set to MAX_INT), or the expected `IllegalArgumentException` will be caught.
```java
@Override
      public void onSuccess(ByteBuffer response) {
        try {
          int size = response.remaining();
          ByteBuffer copy = ByteBuffer.allocate(size); // set size to -1 in runtime when debug
          copy.put(response);
          // flip "copy" to make it readable
          copy.flip();
          result.set(copy);
        } catch (Throwable t) {
          result.setException(t);
        }
      }
```

### CODE REVIEW

**Comment by LantaoJin** [comment]
cc @zsxwing

**Review by srowen — COMMENTED** [review]
Seems OK to me; are there other callbacks that might have the same problem?

**Comment by LantaoJin** [comment]
@srowen , I only find one place in `ExternalShuffleClient.removeBlocks` which maybe has the same problem:
```java
public Future<Integer> removeBlocks(
      String host,
      int port,
      String execId,
      String[] blockIds) throws IOException, InterruptedException {
    checkInit();
    CompletableFuture<Integer> numRemovedBlocksFuture = new CompletableFuture<>();
    ByteBuffer removeBlocksMessage = new RemoveBlocks(appId, execId, blockIds).toByteBuffer();
    final TransportClient client = clientFactory.createClient(host, port);
    client.sendRpc(removeBlocksMessage, new RpcResponseCallback() {
      @Override
      public void onSuccess(ByteBuffer response) {
        BlockTransferMessage msgObj = BlockTransferMessage.Decoder.fromByteBuffer(response);
        numRemovedBlocksFuture.complete(((BlocksRemoved)msgObj).numRemovedBlocks);
        client.close();
      }
```
I prefer to change to below code since `fromByteBuffer` could throw `IllegalArgumentException`
```java
      @Override
      public void onSuccess(ByteBuffer response) {
        try {
          BlockTransferMessage msgObj = BlockTransferMessage.Decoder.fromByteBuffer(response);
          numRemovedBlocksFuture.complete(((BlocksRemoved) msgObj).numRemovedBlocks);
        } catch (Exception e) {
          logger.warn("Error trying to remove RDD blocks " + Arrays.toString(blockIds) +
            " via external shuffle service from executor: " + execId, e);
          numRemovedBlocksFuture.complete(0);
        } finally {
          client.close();
        }
      }
```

**Comment by LantaoJin** [comment]
Should I fix above code if needed in this PR or file a new one?

**Comment by srowen** [comment]
I think you can fix the similar issue here, and update the title/description.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java [hunk 1]
```diff
@@ -163,9 +163,16 @@ public void registerWithShuffleServer(
     client.sendRpc(removeBlocksMessage, new RpcResponseCallback() {
       @Override
       public void onSuccess(ByteBuffer response) {
-        BlockTransferMessage msgObj = BlockTransferMessage.Decoder.fromByteBuffer(response);
-        numRemovedBlocksFuture.complete(((BlocksRemoved)msgObj).numRemovedBlocks);
-        client.close();
+        try {
+          BlockTransferMessage msgObj = BlockTransferMessage.Decoder.fromByteBuffer(response);
+          numRemovedBlocksFuture.complete(((BlocksRemoved) msgObj).numRemovedBlocks);
+        } catch (Exception e) {
```
Comment (srowen) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java]: Do you want to catch `Throwable` here?

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClient.java [hunk 2]
```diff
@@ -237,11 +237,15 @@ public ByteBuffer sendRpcSync(ByteBuffer message, long timeoutMs) {
...
-        // flip "copy" to make it readable
-        copy.flip();
-        result.set(copy);
+        try {
+          ByteBuffer copy = ByteBuffer.allocate(response.remaining());
+          copy.put(response);
+          // flip "copy" to make it readable
+          copy.flip();
+          result.set(copy);
+        } catch (Throwable t) {
```
Comment (srowen) [inline common/network-common/src/main/java/org/apache/spark/network/client/TransportClient.java]: Do you want to log the throwable here just for completeness?

**Review by srowen — CHANGES_REQUESTED** [review]

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java (hunk 1 above)
Comment (LantaoJin) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java]: I haven’t seen any error could occur here, so I use Exception. Throwable of course is fine.

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClient.java (hunk 2 above)
Comment (LantaoJin) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClient.java]: I will add a warning log.

**Comment by LantaoJin** [comment]
Gentle ping @srowen

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master/2.4/2.3

**Pull request closed without merging**